### PR TITLE
[AccessControl] Add Access Control checks to event management

### LINKED ISSUE #14435: Add access control check for IM read events
Need to add a call to AccessControl::Check before creating any event report. Basically, do for read event what has already been done for read attribute.

### PR DESCRIPTION
**Pull request by erjiaqing**
#### Problem
Fixes #14435 

#### Change overview
- Add code for checking ACL when fetching events

#### Testing
@mlepage-google will add test cases and fix potential bugs.

### CODE REVIEW

**Review by woody-apple — APPROVED** [review]

File: src/app/EventManagement.cpp [hunk 1]
```diff
@@ -590,46 +571,115 @@ CHIP_ERROR EventManagement::CopyEvent(const TLVReader & aReader, TLVWriter & aWr
...
+    ReturnErrorOnFailure(builder.CreateErrorStatus().EncodeStatusIB(aStatus).GetError());
+
+    ReturnErrorOnFailure(builder.EndOfEventStatusIB().GetError());
+
+    ReturnErrorOnFailure(aWriter.EndContainer(containerType));
+    ReturnErrorOnFailure(aWriter.Finalize());
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR EventManagement::IsInterestedEventPaths(EventLoadOutContext * eventLoadOutContext,
```
Comment (bzbarsky-apple) [inline src/app/EventManagement.cpp:598]: This function name doesn't really make sense for a function that does not return a boolean....

File: src/app/EventManagement.h [hunk 2]
```diff
@@ -360,7 +362,39 @@ class EventManagement
...
     void SetScheduledEventInfo(EventNumber & aEventNumber, uint32_t & aInitialWrittenEventBytes);
 
+    enum class BypassACL : uint8_t
+    {
+        kNoBypass   = 0,
+        kAlwaysPass = 1,
+        kAlwaysFail = 2,
+    };
+
+    void SetBypassACL(BypassACL aConfig) { mBypassACL = aConfig; }
```
Comment (bzbarsky-apple) [inline src/app/EventManagement.h:372]: Is this only needed for tests?  We should not be shipping this in release software; it increases attack surface...

File: src/app/tests/TestReadInteraction.cpp [hunk 3]
```diff
@@ -986,6 +999,88 @@ void TestReadInteraction::TestReadRoundtripWithSameDifferentPathsDataVersionFilt
...
+        MockInteractionModelApp delegate;
+        NL_TEST_ASSERT(apSuite, !delegate.mGotEventResponse);
+
+        app::ReadClient readClient(chip::app::InteractionModelEngine::GetInstance(), &ctx.GetExchangeManager(), delegate,
+                                   chip::app::ReadClient::InteractionType::Read);
+
+        err = readClient.SendRequest(readPrepareParams);
+        NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+
+        InteractionModelEngine::GetInstance()->GetReportingEngine().Run();
```
Comment (bzbarsky-apple) [inline src/app/tests/TestReadInteraction.cpp:1040]: Please don't make manual `Run()` calls.  Those should be happening automatically as long as we run the event loop (which we should).  If we make manual `Run` calls in tests, that just covers up bugs in Run scheduling.

File: src/app/tests/TestReadInteraction.cpp [hunk 4]
```diff
@@ -986,6 +999,88 @@ void TestReadInteraction::TestReadRoundtripWithSameDifferentPathsDataVersionFilt
...
+        NL_TEST_ASSERT(apSuite, delegate.mGotEventResponse);
+        NL_TEST_ASSERT(apSuite, delegate.mNumReadEventFailureStatusReceived > 0);
+        NL_TEST_ASSERT(apSuite, delegate.mLastStatusReceived.mStatus == Protocols::InteractionModel::Status::UnsupportedAccess);
+        NL_TEST_ASSERT(apSuite, !delegate.mReadError);
+    }
+
+    GenerateEvents(apSuite, apContext);
+
+    EventManagement::GetInstance().SetBypassACL(chip::app::EventManagement::BypassACL::kAlwaysFail);
+    // When read events with withcard paths without enough privilege for reading one or more event, we will not exclude the events
```
Comment (bzbarsky-apple) [inline src/app/tests/TestReadInteraction.cpp:1050]:
```suggestion
    // When reading events with withcard paths without enough privilege for reading one or more event, we will not exclude the events
```
I don't understand that comment.  We _should_ be excluding those events....

File: src/app/tests/TestReadInteraction.cpp [hunk 5]
```diff
@@ -986,6 +999,88 @@ void TestReadInteraction::TestReadRoundtripWithSameDifferentPathsDataVersionFilt
...
+        MockInteractionModelApp delegate;
+        NL_TEST_ASSERT(apSuite, !delegate.mGotEventResponse);
+
+        app::ReadClient readClient(chip::app::InteractionModelEngine::GetInstance(), &ctx.GetExchangeManager(), delegate,
+                                   chip::app::ReadClient::InteractionType::Read);
+
+        err = readClient.SendRequest(readPrepareParams);
+        NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+
+        InteractionModelEngine::GetInstance()->GetReportingEngine().Run();
```
Comment (bzbarsky-apple) [inline src/app/tests/TestReadInteraction.cpp:1071]: Again, please no manual Run calls.

File: src/controller/TypedReadCallback.h [hunk 6]
```diff
@@ -149,6 +149,8 @@ class TypedReadEventCallback final : public app::ReadClient::Callback
         CHIP_ERROR err = CHIP_NO_ERROR;
         DecodableEventType value;
 
+        VerifyOrExit(apStatus != nullptr && !apStatus->IsSuccess(), err = apStatus->ToChipError());
```
Comment (bzbarsky-apple) [inline src/controller/TypedReadCallback.h:152]: This will crash if `apStatus` is null.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: src/app/EventManagement.h (hunk 2 above)
Comment (yunhanw-google) [reply src/app/EventManagement.h:372]: agree, don't put it in release software, consider to use CONFIG_IM_BUILD_FOR_UNIT_TEST?

File: src/controller/TypedReadCallback.h (hunk 6 above)
Comment (erjiaqing) [reply src/controller/TypedReadCallback.h:152]: The condition is wrong, it should be negatived -- `apStatus == nullptr || apStatus->IsSuccess()`

File: src/app/tests/TestReadInteraction.cpp (hunk 4 above)
Comment (erjiaqing) [reply src/app/tests/TestReadInteraction.cpp:1050]: `will not exclude` -> `will exclude`

File: src/app/EventManagement.cpp (hunk 1 above)
Comment (erjiaqing) [reply src/app/EventManagement.cpp]: Renamed to `CheckEventContext` with updated return values.

File: src/app/tests/TestReadInteraction.cpp (hunk 3 above)
Comment (erjiaqing) [reply src/app/tests/TestReadInteraction.cpp:1040]: We don't have a function for one-shot event processing for now.

We have a comment for 
`This DOES NOT ensure that all pending events are serviced to completion (i.e timers, any ScheduleWork calls).`

In `DrainAndServiceIO`

But invoking works scheduled by `ScheduleWork` is what we want here.

This should be a follow up as "Support one-shot event dispatch for tests"

Comment (bzbarsky-apple) [reply src/app/tests/TestReadInteraction.cpp:1040]: > But invoking works scheduled by ScheduleWork is what we want here.

You may want to update to tip and read the updated documentation for DrainAndServiceIO.

Again, we should not be adding `Run` calls.

File: src/app/tests/TestReadInteraction.cpp (hunk 5 above)
Comment (yunhanw-google) [reply src/app/tests/TestReadInteraction.cpp:1071]: manual run call has been removed in tests in master branch, consider to rebase it. @erjiaqing

File: src/app/EventManagement.h (hunk 2 above)
Comment (mlepage-google) [reply src/app/EventManagement.h]: Please coordinate with me about Access Control during testing. Yes, it's a concern I'm trying to keep under control. Don't add bypass flags or compilation flags.

Instead, in a test access control can be bypassed by configuring a different AccessControl::Delegate (implementation), such as the new "permissive" one. Alternatively, in test code an "admin for CASE subject" ACL can be installed, then even other event listeners (not just the commissioner or admin) can succeed in IM operations during tests. There's issues here with leaving the ACL in a good state, so other tests can succeed (e.g. on devices where all tests run in one app).

I know this isn't ideal, but those are the approaches in the short term, until I can get cycles to improve it more (which will eventually rise to the top of my priority list, but isn't now).

**Comment by mlepage-google** [comment]
Terminology wise, please change "ACL check" to "Access Control check" where it occurs. Access can be granted without an ACL. I only actually use the term ACL (more specifically, ACL entry) for the data thing that is used as part of those checks. Thanks.

File: src/app/EventManagement.h (hunk 2 above)
Comment (mlepage-google) [reply src/app/EventManagement.h]: Just looking at this PR in more detail, it seems two unit test files were affected. I did add code in AppTestContext.cpp which sets the "permissive" access control implementation, so these app level tests (TestCommandInteraction, for example) should pass. So you shouldn't need another bypass. If you do, please let me know what is failing/complaining, so I can address it or at least provide a recommendation for how to address it. Thanks!

File: src/app/EventManagement.cpp [hunk 7]
```diff
@@ -590,46 +571,118 @@ CHIP_ERROR EventManagement::CopyEvent(const TLVReader & aReader, TLVWriter & aWr
...
+    if (aclError != CHIP_NO_ERROR)
+    {
+        ReturnErrorCodeIf(aclError != CHIP_ERROR_ACCESS_DENIED, aclError);
+        if (isPathInterestedByConcretePath)
+        {
+            ret = CHIP_ERROR_ACCESS_DENIED;
+        }
+        else
+        {
+            ret = CHIP_ERROR_UNEXPECTED_EVENT;
```
Comment (mlepage-google) [inline src/app/EventManagement.cpp:662]: Is this more like "unexpected error" or actually meaning that the event (being processed) is not expected?

Comment (erjiaqing) [reply src/app/EventManagement.cpp:662]: This means the event being processed is not interested

**Review by bzbarsky-apple — APPROVED** [review]

**Review by yunhanw-google — APPROVED** [review]

File: src/app/EventManagement.h [hunk 8]
```diff
@@ -449,11 +486,26 @@ class EventManagement
         return CHIP_ERROR_NO_MEMORY;
     };
 
+    /**
+     * @brief checking if the given path is an interested path, this is a helper function for EventManagement::EventIterator
+     *
+     * @retval CHIP_ERROR_UNEXPECTED_EVENT This path should be excluded in the generated event report.
+     * @retval CHIP_EVENT_ID_FOUND This path should be included in the generated event report.
```
[the remark this answers is not in the document]
Comment (erjiaqing) [reply src/app/EventManagement.h:493]: #15713

File: src/app/EventManagement.h (hunk 2 above)
Comment (erjiaqing) [reply src/app/EventManagement.h]: Commented out the test with a TODO for #10253

**Pull request merged**